Add RoomName struct to ruma_events::room::name.

Frinksy opened the pull request:
Resolves: https://github.com/ruma/ruma/issues/159

I'm not sure if I'm going in the right direction for this PR. I've changed the constructor's signature for `NameEventContent` which would be a breaking change.

Also, I'm not sure if creating a `RoomName` type should fail if the name is an empty `String`.
There is one test which fails, but that is due to the fact that currently it's possible to create an empty `RoomName`.

Frinksy committed:
Add RoomName struct to ruma_events::room::name.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -14,7 +16,7 @@ pub type NameEvent = StateEvent<NameEventContent>;
     /// The name of the room. This MUST NOT exceed 255 bytes.
     #[serde(default, deserialize_with = "room_name")]
-    pub(crate) name: Option<String>,
+    pub(crate) name: Option<RoomName>,
```
With the type now enforcing the character limitation, the field can be made public and the `MUST NOT exceed` part of the docs removed.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -23,21 +25,20 @@ impl NameEventContent {
     /// # Errors
     ///
     /// `InvalidInput` will be returned if the name is more than 255 bytes.
```
This is now outdated.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -23,21 +25,20 @@ impl NameEventContent {
+        match &self.name {
+            Some(_) => Some(self.name.as_ref().unwrap().0.as_str()),
+            None => None,
+        }
```
This is a very convoluted way to write this. `self.name.as_ref().map(|r| r.0.as_str())` would work too, but really I think this should just return `&self.name` and have the signature updated accordingly.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -14,7 +16,7 @@ pub type NameEvent = StateEvent<NameEventContent>;
 pub struct NameEventContent {
     /// The name of the room. This MUST NOT exceed 255 bytes.
     #[serde(default, deserialize_with = "room_name")]
```
This should now be `deserialize_with = "ruma_serde::empty_string_as_none"`.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -47,15 +48,43 @@ where
+/// The name of a room.
+///
+/// It should not exceed 255 characters.
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
```
I think it might be good to have `#[serde(transparent)]` here for the derived `Serialize` impl (it should then be a bit simpler, even though the outcome will be the same for JSON).

The `Deserialize` impl should not be derived, otherwise it's possible to create an empty or too long `RoomName` that way.

jplatte reviewed (commented):
Yes, this was definitely expected to be a breaking change. I'll update the base branch accordingly. In the future, please select the base branch to be `next` for breaking changes (it might also be beneficial to use that as the starting point of your branch).

Re. empty `RoomName`, I think that should be disallowed. The serde attribute on the field will prevent that from being a deserialization error in that context, but being able to have an empty `RoomName` instance seems not useful.

Frinksy commented:
I've updated the target branch and rebased my branch onto `next`.

Also, I've added an `#[allow(clippy::exhaustive_structs)]` to `RoomName` since I expect that it will always be just a string.

DevinR528 commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -11,51 +13,71 @@ pub type NameEvent = StateEvent<NameEventContent>;
-    pub fn name(&self) -> Option<&str> {
-        self.name.as_deref()
+    pub fn name(&self) -> &Option<RoomName> {
+        &self.name
```
```suggestion
        self.name.as_deref()
```

I think `Option<&RoomName>` would be more consistent?

DevinR528 commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -11,51 +13,71 @@ pub type NameEvent = StateEvent<NameEventContent>;
 /// The payload for `NameEvent`.
 #[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
 #[ruma_event(type = "m.room.name", kind = State)]
+#[non_exhaustive]
```
```suggestion
#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
```

Unless there is a reason to not have this configurable?

DevinR528 commented:
> Also, I've added an #[allow(clippy::exhaustive_structs)] to RoomName since I expect that it will always be just a string.

It also makes it really awkward to use `RoomName { 0: raw_string, .. }` not that there would be any point in doing that because of all the convenience methods and such. But yeah, I don't think there is any tuple struct with a `non_exhaustive` attr.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -11,51 +13,71 @@ pub type NameEvent = StateEvent<NameEventContent>;
+/// It should not exceed 255 characters and should not be empty.
+#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
+#[serde(transparent)]
+#[allow(clippy::exhaustive_structs)]
```
I'm pretty sure this can be removed, as the lint only triggers on structs with all-public fields in the first place.

Frinksy replied:
Yes, I think I'd originally just misread one of the clippy messages. I've removed it now.

Frinksy committed:
Change name accessor and remove clippy bypass.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -11,51 +13,78 @@ pub type NameEvent = StateEvent<NameEventContent>;
     /// The name of the room, if any.
-    pub fn name(&self) -> Option<&str> {
+    pub fn name(&self) -> Option<&RoomName> {
         self.name.as_deref()
```
This can use the "regular" `as_ref`, then the no-op Deref impl can be removed or changed to `Target = str`.

Frinksy committed:
Use Target = str in RoomName Deref implementation.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -11,51 +13,78 @@ pub type NameEvent = StateEvent<NameEventContent>;
+    }
+}
+
+impl Deref for RoomName {
```
Thinking more about it, since we don't have `Deref` impls for the types in ruma-identifiers, let's not have on here either for now.

jplatte commented on crates/ruma-events/src/room/name.rs:
```diff
@@ -64,12 +93,14 @@ mod tests {
         let name_event = StateEvent {
-            content: NameEventContent { name: Some("The room name".into()) },
+            content: NameEventContent {
+                name: RoomName::try_from("The room name".to_string()).ok(),
```
This is a personal pet peeve of mine.. See also clippy lint [str_to_string](https://rust-lang.github.io/rust-clippy/master/index.html#str_to_string).

```suggestion
                name: RoomName::try_from("The room name".to_owned()).ok(),
```

Frinksy committed:
Remove Deref trait for RoomName and apply style changes.

jplatte commented:
Thanks!
